specs/IFramePermissionRequested.md

```diff
@@ -0,0 +1,363 @@
+                                    }
+                                    else if (response == IDNO)
+                                    {
+                                        m_cachedPermissions[cachedKey] = false;
+                                        state = COREWEBVIEW2_PERMISSION_STATE_DENY;
+                                    }
+
+                                    CHECK_FAILURE(args->put_State(state));
+
+                                    PutHandled(args);
+                                    return S_OK;
+                            }).Get(),
+                            &m_PermissionRequestedToken));
+                    }
+                    return S_OK;
+            }).Get(),
+            &m_FrameCreatedToken));
+    }
+}
+
+static PCWSTR NameOfPermissionKind(COREWEBVIEW2_PERMISSION_KIND kind)
+{
+    switch (kind)
+    {
+        case COREWEBVIEW2_PERMISSION_KIND_MICROPHONE:
+            return L"Microphone";
+        case COREWEBVIEW2_PERMISSION_KIND_CAMERA:
+            return L"Camera";
+        case COREWEBVIEW2_PERMISSION_KIND_GEOLOCATION:
+            return L"Geolocation";
+        case COREWEBVIEW2_PERMISSION_KIND_NOTIFICATIONS:
+            return L"Notifications";
+        case COREWEBVIEW2_PERMISSION_KIND_OTHER_SENSORS:
+            return L"Generic Sensors";
+        case COREWEBVIEW2_PERMISSION_KIND_CLIPBOARD_READ:
+            return L"Clipboard Read";
+        default:
+            return L"Unknown resources";
+    }
+}
+
+static void PutHandled(ICoreWebView2PermissionRequestedEventArgs* args)
+{
+    // In the case of an iframe requesting permission, the default behavior is
+    // to first raise the PermissionRequested event off of the CoreWebView2Frame
+    // and invoke it's handlers, and then raise the event off the CoreWebView2
+    // and invoke it's handlers. However, If we set Handled to true on the
+    // CoreWebView2Frame event handler, then we will not raise the
+    // PermissionRequested event off the CoreWebView2.
+    wil::com_ptr<ICoreWebView2PermissionRequestedEventArgs2> args2;
+    CHECK_FAILURE(args->QueryInterface(IID_PPV_ARGS(&args2)));
+    if (args2)
+    {
+        CHECK_FAILURE(args2->put_Handled(true));
+    }
+}
+```
+
+## C#: Registering IFrame Permission Requested Handler
+```c#
+private WebView2 m_webview;
+Dictionary<Tuple<string, CoreWebView2PermissionKind, bool>, bool> m_cachedPermissions;
+
+// Example Use Case - an app may want to specify that all iframes should
+// create a message box to prompt the user for approval on a permission request.
+// The approval state could be cached so that future requests are automatically
+// handled if they have been requested previously.
+void RegisterIFramePermissionRequestedHandler()
+{
+    m_webview.CoreWebView2.FrameCreated += (sender, frameCreatedArgs) =>
+    {
+        frameCreatedArgs.Frame.PermissionRequested += (frameSender, permissionArgs) =>
```

Comment:
In the C++ case you check if the CoreWebView2Frame implements the new interface that has PermissionRequested on it. In C#/WinRT doing Frame.PermissionRequested += will do the same thing under the covers but will throw if the interface isn't implemented. To match the C++ code you'd need a try/catch and ignore the exception with a comment explaining why.